### PR TITLE
Carl ansi part1

### PR DESCRIPTION
**Pull request by ccodere**
The last important part for ANSI C conversion. Now libobject also uses ANSI C standard I/O as well as some of the utilities in amisc when feasible.

The important changes here are found in the files:
object.h 
archiver.c

I have one important doubt that you can find in comm4.c line 246, the old code would write beyond the array size, but if i am not sure if my fix is correct. 

Please let me know if you find any issues. Tests all pass on my system.

### CODE REVIEW

**Comment by ccodere** [comment]
Let me investigate the issue ...

**Comment by davidgiven** [comment]
I agree with you about comm4.c. That looks dead wrong --- the array's only 14 bytes long, so ar_hdr[14] must be out of bounds. OTOH it always was wrong, so this is unlikely to be the cause of the current build failure on Linux. I can duplicate that on amd64 but not on ARM, so this could well be some sort of sizeof(ptr) memory corruption thing.

**Comment by ccodere** [comment]
> I agree with you about comm4.c. That looks dead wrong --- the array's only 14 bytes long, so 

ok, I will first investigate my issue first then go to next one... it may take a while, as i discovered that even without my changes on my linux lubuntu system, newdispose_p test completely freezes the build... 

I also tested with address sanitizer, and i discovered at least 2 existing memory overflows - ... (already fixed locally)... 

But i need to test a lot more before i even try to push anything else...

**Comment by davidgiven** [comment]
I got some useful stack traces out of valgrind in memory debugger mode:

```
==3290== Invalid write of size 1
==3290==    at 0x4838DD7: strcpy (vg_replace_strmem.c:512)
==3290==    by 0x10EBC6: innumber (comm5.c:472)
==3290==    by 0x10E08F: yylex (comm5.c:37)
==3290==    by 0x11110E: yyparse (y.tab.c:2889)
==3290==    by 0x10D80C: parse (comm4.c:328)
==3290==    by 0x10D722: pass_1 (comm4.c:183)
==3290==    by 0x10D5D0: main (comm4.c:117)
==3290==  Address 0x4a51623 is 0 bytes after a block of size 3 alloc'd
==3290==    at 0x48356AF: malloc (vg_replace_malloc.c:298)
==3290==    by 0x4837DE7: realloc (vg_replace_malloc.c:826)
==3290==    by 0x10EBA9: innumber (comm5.c:470)
==3290==    by 0x10E08F: yylex (comm5.c:37)
==3290==    by 0x11110E: yyparse (y.tab.c:2889)
==3290==    by 0x10D80C: parse (comm4.c:328)
==3290==    by 0x10D722: pass_1 (comm4.c:183)
==3290==    by 0x10D5D0: main (comm4.c:117)
```

It looks like they're all the same, which suggests the string allocation routine at line 470 of comm5.c is mssing a +1 somewhere. Er, I think I wrote that?

**Comment by kernigh** [comment]
There is a pre-existing bug. I tried to build the default branch (not this pull request) with `MALLOC_OPTIONS=S gmake` in OpenBSD/amd64, got 2 "chunk canary" errors like

```
[2087/9210] ACKDIR=/home/kernigh/park/.../cem/libcc.ansi/core/math/sqrt.c  -O6 
FAILED: /home/kernigh/park/ack-build/obj/lang/cem/libcc.ansi/lib_linuxmips/sqrt$
sqrt.o                                                                         
ACKDIR=/home/kernigh/park/ack-build/staging /home/kernigh/park/ack-build/stagin$
/bin/ack -mlinuxmips -c.o -o /home/kernigh/park/ack-build/obj/lang/cem/libcc.an$
i/lib_linuxmips/sqrt/sqrt.o lang/cem/libcc.ansi/core/math/sqrt.c  -O6          
as(21814) in realloc(): chunk canary corrupted 0x39123ad5500 0x3@0x3           
/home/kernigh/park/ack-build/staging/bin/ack: /home/kernigh/park/ack-build/stag$
ng/lib/ack/linuxmips/as died with signal 6       
```

**Comment by kernigh** [comment]
If the assembler encountered a floating-point constant before the first string constant, it allocated stringlen bytes but copied stringlen + 1 bytes (including the '\0'). The below diff prevented my "chunk canary corrupted" error. @ccodere, you might add something like this to the pull request. (I forgot the void in `need_stringbuf(void)`.)

```diff
diff --git a/mach/proto/as/comm5.c b/mach/proto/as/comm5.c
index 0b448ef0e..612f7b328 100644
--- a/mach/proto/as/comm5.c
+++ b/mach/proto/as/comm5.c
@@ -395,6 +395,18 @@ char* readident(int c)
 }
 #endif
 
+static void need_stringbuf()
+{
+	if (!maxstring)
+	{
+		maxstring = STRINGMAX;
+		if ((stringbuf = malloc(maxstring)) == 0)
+		{
+			fatal("out of memory");
+		}
+	}
+}
+
 static int innumber(int c)
 {
 	char* p;
@@ -464,11 +476,8 @@ floatconstant:
 
 	*p = '\0';
 	stringlen = p - num;
-	if (stringlen > maxstring)
-	{
-		maxstring = stringlen;
-		stringbuf = realloc(stringbuf, maxstring);
-	}
+	need_stringbuf();
+	assert(stringlen < maxstring);
 	strcpy(stringbuf, num);
 	return NUMBERF;
 }
@@ -478,14 +487,7 @@ static int instring(int termc)
 	char* p;
 	int c;
 
-	if (!maxstring)
-	{
-		maxstring = STRINGMAX;
-		if ((stringbuf = malloc(maxstring)) == 0)
-		{
-			fatal("out of memory");
-		}
-	}
+	need_stringbuf();
 	p = stringbuf;
 	for (;;)
 	{
```

**Comment by ccodere** [comment]
> If the assembler encountered a floating-point constant before the first string constant, it allocated >stringlen bytes but copied stringlen + 1 bytes (including the '\0'). The below diff prevented my "chunk c

Thanks, i will patch and try to push if all my other changes are also ok ...

I will also commit a few other memory fixes that i discovered.

**Comment by ccodere** [comment]
I have done the patch proposed by kernigh on my local machine, but i still have the newdispose test freezing, i will look into it tomorrow.

**Comment by ccodere** [comment]
Finally everything is fixed, thanks for all your support and help. The tests were not passing on my linux machine because of a buggy qemu-i386, so the newdispose_p does work as expected after all. I did fix some other memory overflows though.

**Comment by ccodere** [comment]
More updates for ANSI C clean compilation, also updated most man pages. Still left to do some conversion in some of the utilities, including print library.

**Comment by davidgiven** [comment]
Thanks for this --- but I'm actually off on holiday for a week starting in about 8 hours, so I'll probably delayed getting back to you. Sorry!

**Comment by ccodere** [comment]
> Thanks for this --- but I'm actually off on holiday for a week starting in about 8 hours, so I'll probably delayed getting back to you. Sorry!

It is ok it gives me time to fix other things..

File: mach/proto/as/comm4.c [hunk 1]
```diff
@@ -243,8 +244,8 @@ archive(void) {
 		if (needed()) {
 			fseek(input,offset,0);
 			archsize = header.ar_size;
-			header.ar_name[14] = '\0';
-			parse(remember(header.ar_name));
+			ar_name[AR_NAME_MAX] = '\0';
```
Comment (davidgiven) [inline mach/proto/as/comm4.c]: Buffer overflow; this should be AR_NAME_MAX-1 (or ar_hdr needs to define the field as char[AR_NAME_MAX+1]. Looks like this bug has always been here. I've seen similar bugs before.

Comment (davidgiven) [inline mach/proto/as/comm4.c:247]: Nice catch. Looks like that bug's been there for at least 33 years.

**Comment by davidgiven** [comment]
Sorry this dropped off my radar --- thank you very much for doing all this; it's a tonne of work. I spotted one typo in a comment but it's not worth delaying merging for.

Incidentally, this PR contains multiple unrelated changes; these would be much better split up into different PRs. Smaller PRs are so much easier (and therefore faster) to review, and make for a clearer revision history. Ideally each logical change should be a single PR. I've noticed you've filed a bunch of bugs recently (which is great!). Any chance we could get a PR for each one?

Thanks!

**Pull request merged**